File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    
     int sheep = 1;
     int wolf = 0;
    int answer =0;
    public int solution(int[] info, int[][] edges) {
   
        int n = info.length;
        ArrayList<Integer> []tree  = new ArrayList[n];
        
        for (int i= 0; i< n; i++) {
            tree[i] = new ArrayList<>();
        }
        for (int [] temp : edges) {
            int start = temp[0];
            int end = temp[1];
            tree[start].add(end);
        }
        ArrayList<Integer> candidates = new ArrayList<>();
        candidates.addAll(tree[0]);
        
        dfs(1,0,candidates, tree, info);
        
        return answer;
    }
    
    void dfs (int sheep, int wolf, ArrayList<Integer> candidates, ArrayList<Integer> [] tree, int[]info) {
        
        answer = Math.max(answer, sheep);
        
        for (int nxt : candidates) {
            int next_sheep = sheep;
            int next_wolf = wolf;
            
            if (info[nxt] == 1)
                next_wolf +=1;
            
            if (info[nxt]==0)
                next_sheep +=1;
            
            if (next_sheep <= next_wolf) 
                continue;
            
            ArrayList<Integer> next_candidates = new ArrayList<>(candidates);
            next_candidates.remove(Integer.valueOf(nxt));
            
            for (int nxt_node : tree[nxt]) {
                next_candidates.add(nxt_node);
            } 
            
            dfs (next_sheep, next_wolf, next_candidates, tree, info);
        }
        
        
    }
    
    
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
def solution(info, edges):
    answer = 0
    
#     n = len(info)
    
#     tree = [[]*(n) for _ in range (n)]
#     for e in edges :
#         start,end = e
#         tree[start].append(end)

#     def dfs(sheep, wolf,candidates):
#         nonlocal answer
#         answer= max(answer, sheep)

#         for nxt in candidates:
#             next_sheep=sheep
#             next_wolf =wolf
        
#             if info[nxt] == 0 :
#                 next_sheep +=1
            
#             if info[nxt] ==1 :
#                 next_wolf +=1
                    
#             if next_sheep <= next_wolf :
#                 continue
                
#             next_candidates = candidates.copy()
#             next_candidates.remove(nxt)
#             next_candidates.extend(tree[nxt])
#             dfs(next_sheep, next_wolf, next_candidates)
    
#     dfs(1, 0, tree[0])
    
#     return answer
            
    n = len(info)
    tree = [[] *(n) for _ in range (n)]
    
    for e in edges :
        a = e[0]
        b = e[1]
        tree[a].append(b)
        
    def dfs (sheep, wolf, candidates) :
        nonlocal answer
        answer = max(answer, sheep)

        for nxt_node in candidates:

            next_sheep = sheep
            next_wolf = wolf
        
            
            if info[nxt_node] == 0 :
                next_sheep+=1
            
            if info[nxt_node] == 1:
                next_wolf +=1
            
            if next_sheep <= next_wolf :
                continue
            
            next_candidates = candidates.copy()
            next_candidates.remove(nxt_node)
            next_candidates.extend(tree[nxt_node])
            dfs(next_sheep, next_wolf, next_candidates)
    
            
            
        
        
    dfs(1,0, tree[0])    
    
        
    return answer

        
        